runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java

```diff
@@ -382,17 +382,24 @@ private void syncObjectUnsafeInner(long timestamp) {
      */
     public void setUncommittedChanges() {
         WriteSetSMRStream stream = getOptimisticStreamUnsafe();
-        if (stream == null
-                || !stream.isStreamCurrentContextThreadCurrentContext()) {
 
-            // We are setting the current context to the root context of nested transactions.
-            // Upon sync forward
-            // the stream will replay every entries from all parent transactional context.
-            WriteSetSMRStream newSmrStream =
-                    new WriteSetSMRStream(TransactionalContext.getTransactionStackAsList(), getID());
+        // Nothing to do, the stream and the thread current contexts are the same.
+        if (stream != null && stream.isStreamCurrentContextThreadCurrentContext()) {
+            return;
+        }
 
-            setUncommittedChanges(newSmrStream);
+        // Monotonic objects do not allow for read-your-writes within a transaction.
+        if (isMonotonicObject()) {
+            return;
         }
+
+        // We are setting the current context to the root context of nested transactions.
+        // Upon sync forward
+        // the stream will replay every entries from all parent transactional context.
+        WriteSetSMRStream newSmrStream =
+                new WriteSetSMRStream(TransactionalContext.getTransactionStackAsList(), getID());
+
+        setUncommittedChanges(newSmrStream);
```

Comment:
You can extend `WriteSetSMRStream` and make the stream operations emit "no-ops", that is, empty collections for the streaming API (e.g., remainingUpTo, current, previous and pos). That way monotonic objects don't require special handling. 

Although that might be more boilerplate code, I think it's cleaner and simpler to reason about since there is no special handling for monotonic objects. Sine the no-op stream never produces any state there the existing VLO machinery would work AND it will produce less garbage.

Reply:
Good idea, however, the issue is that we need to set the upcall result. If you make the `WriteSetSMRStream` return an empty collection, the upcall result will never be set. You will end up with a transaction abort:

```
org.corfudb.runtime.exceptions.TransactionAbortedException: TX ABORT  | Snapshot Time = Token(epoch=0, sequence=-1) | Failed Transaction ID = a564c94c-c586-c8de-17f2-d66e0907ec3a | Offending Address = -1 | Conflict Key = 00 | Conflict Stream = 1889b148-5a6c-3a55-9369-060f3f230293 | Cause = UNDEFINED | Time = 0 ms | Message = Tried to get upcall during a transaction but we don't have it even after an optimistic sync (asked for 0 we have 0-0)
```

Reply:
That is because a MutatorAccessor is being used and it assumes that the Mutator must to produce a mutation. Since monotonic objects are not MutatorAccessors, but its annotated as such to implement the map interface you can allow the `getUpcallResult` to deal with a no-op mutator, essentially it will return null as opposed to throwing an exception.

Reply:
I suppose the less complicated the VLO is, the better. I stubbed `getUpcallResultInner` for monotonic objects that are within transaction. @Maithem can you take a look at it again?